syntax = "proto3";

package post;

service PostService {
  rpc CreatePost(CreatePostRequest) returns (PostResponse) {}
  rpc GetPost(GetPostRequest) returns (PostResponse) {}
  rpc UpdatePost(UpdatePostRequest) returns (PostResponse) {}
  rpc DeletePost(DeletePostRequest) returns (DeletePostResponse) {}
  rpc ListPosts(ListPostsRequest) returns (ListPostsResponse) {}
}

message Post {
  int64 id = 1;
  string title = 2;
  string description = 3;
  string username = 4;
  string created_at = 5;
  string updated_at = 6;
  bool is_private = 7;
  repeated string tags = 8;
}

message CreatePostRequest {
  string title = 1;
  string description = 2;
  string username = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message GetPostRequest {
  int64 post_id = 1;
  string username = 2;
}

message UpdatePostRequest {
  int64 post_id = 1;
  string username = 2;
  string title = 3;
  string description = 4;
  bool is_private = 5;
  repeated string tags = 6;
}

message DeletePostRequest {
  int64 post_id = 1;
  string username = 2;
}

message DeletePostResponse {
  bool success = 1;
  string message = 2;
}

message ListPostsRequest {
  int64 page = 1;
  int64 page_size = 2;
  string username = 3;
  string tag = 4;
}

message ListPostsResponse {
  repeated Post posts = 1;
  int64 total = 2;
  int64 page = 3;
  int64 page_size = 4;
  int64 pages = 5;
}

message PostResponse {
  Post post = 1;
}
